kernel/src/memory/physical_allocator.rs

```diff
@@ -0,0 +1,207 @@
+///! A modified buddy bitmap allocator. Written originally in
+/// [buddy allocator workshop](https://github.com/Restioson/buddy-allocator-workshop).
+use core::{mem, ptr, ops::Range};
+#[cfg(test)]
+use std::boxed::Box;
+use spin::{Mutex, Once};
+use super::bootstrap_heap::BootstrapHeapBox;
+
+/// Number of orders.
+const LEVEL_COUNT: u8 = 19;
+/// The base order size. All orders are in context of this -- i.e the size of a block of order `k`
+/// is `2^(k + MIN_ORDER)`, not `2^k`.
+const BASE_ORDER: u8 = 12;
+
+/// The physical frame allocator. Requires the bootstrap heap to be initialized, or else the
+/// initializer will panic.
+pub static PHYSICAL_ALLOCATOR: PhysicalAllocator<'static> = PhysicalAllocator {
+    trees: Once::new(),
+    #[cfg(test)]
+    _phantom: ::core::marker::PhantomData,
+};
+
+// Panics from `buddy_allocator.rs` will say they're from here. Go there instead.
+buddy_allocator_bitmap_tree!(LEVEL_COUNT = LEVEL_COUNT, BASE_ORDER = BASE_ORDER);
+
+impl<'a> Tree<BootstrapHeapBox<'a, [Block; BLOCKS_IN_TREE]>> {
+    /// Creates a new tree. Panics if the bootstrap allocator has not been initialized or
+    /// does not have enough memory. Takes a list of usable memory areas, **local to this
+    /// tree.**
+    pub fn new_bootstrap_heap<I>(usable: I) -> Self
+        where I: Iterator<Item=Range<usize>> + Clone,
+    {
+        let mut flat_blocks: BootstrapHeapBox<[Block; BLOCKS_IN_TREE]> = unsafe {
+            super::bootstrap_heap::BOOTSTRAP_HEAP.allocate_zeroed()
+            .expect("No heap memory available!")
+        };
+        Self::new(usable, flat_blocks)
+    }
+}
+
+pub struct PhysicalAllocator<'a> {
```

Comment:
Might be better to make a separate struct for testing.

Reply:
Why? It's just simpler to use cfg's, and we don't have to keep synchronising the testing and the non testing. It's also less code.

Reply:
Fair, though I'd imagine it'd introduce better readability. For example, I got confused why the `tree` field was being declared twice.

Reply:
In this case, imo the complications outweigh the gains. Any way you do it, something will be declared twice.

Reply:
It's less about things being declared twice than it is about separation of concerns. Fair enough, though. Either works.